### PR TITLE
New: Generate open graph tags

### LINKED ISSUE #222: Add opengraph tags to pages
I think it will be nice to add the OG tags to the pages so when people share the urls it looks better.
There should be
* an image: Nellie by default, although it will be nice to be capable of overriding it somehow from the docs.
* title
* description: in the rules' docs maybe the "why is this important" part?

@alrra, @sarvaje @qzhou1607 @ststimac ?

### PR DESCRIPTION
**Pull request by qzhou1607**
## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonar)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/developer-guide/contributing/pull-requests.html#commitmessages)

## Short description of the change(s)

Fix #222

* For rule pages, the content after `Why is this important?` is extracted and used as `description` in the meta tag.
* For other  pages, the content of title is used as `description` for now.
* Every page has a default image, but both the image or description can now be overridden by manually adding front matter in the **original** markdown file in sonar. 
* Remove some unused code.

### CODE REVIEW

**Comment by molant** [comment]
@qzhou1607 can you update this to use actually the text before the "Why is this important?" section?  After reviewing the content in rules I think it's probably more appropriate to use that text.
Instead of squashing can you just do a push? I'll be working on this branch to add the description of the sections that cannot be calculated automatically so that way we will not step in each other (and we can squash once all the changes are done).

Thanks!

**Comment by qzhou1607** [comment]
@molant Sure. By the way I'm working on refactoring some code in `updater.js`, so the code there can be reviewed later.

**Comment by qzhou1607** [comment]
@molant Done with refactoring. Now when the updater runs, it only adds properties that don't already exist in the frontmatter, so all properties in the frontmatter can be overridden.

Also, I just realize that if you need to add description in the frontmatter manually, it needs to happen before the updater runs, so the changes should be in the original markdown files in the sonar repository, right?

**Comment by molant** [comment]
> so the changes should be in the original markdown files in the sonar repository

There are a few files that are in our repo (404, about, contributors, docs index) but for others I'm still trying to figure out what would be the best approach as I'm not very keen on adding frontmatter in that repo. I might add a list of descriptions in `updater.js` and use the right one in each case.

File: src/hexo/themes/sonarwhal/layout/partials/metas.hbs [hunk 1]
```diff
@@ -1,3 +1,6 @@
+<meta property="og:image" content="{{setDefault page.image '/images/sonar-logo.svg'}}" />
```
Comment (alrra) [inline src/hexo/themes/sonarwhal/layout/partials/metas.hbs]: According to [this](https://indieweb.org/The-Open-Graph-protocol#Does_not_support_SVG_images), SVG is not supported.

Comment (molant) [reply src/hexo/themes/sonarwhal/layout/partials/metas.hbs]: I was thinking about using the apple-touch-icon

File: helpers/updater.js [hunk 2]
```diff
@@ -46,10 +54,42 @@ const getLayout = (filePath) => {
     return current;
 };
 
-const generateFrontMatterInfo = (filePath, title) => {
+const trimAndUnquote = (string) => {
+    return string.trim().replace(/^"|"$/, '');
+};
+
+/** Convert frontmatter string to object. */
+const toObeject = (frontMatter) => {
```
Comment (sarvaje) [inline helpers/updater.js]: I think there is a typo `toObeject` should be `toObject` right?

File: helpers/updater.js [hunk 3]
```diff
@@ -46,10 +54,42 @@ const getLayout = (filePath) => {
...
+    let match;
+    const frontmatterObject = {};
+    const frontMatterItemRegex = /(.*):\s*"?(.*)"?/g; // `property: "value"` or `property: value`.
+
+    while ((match = frontMatterItemRegex.exec(frontMatter)) !== null) {
```
Comment (sarvaje) [inline helpers/updater.js]: move the assignation to a different line.
```js
match = frontMatterItemRegex.exec(frontMatter);
while (match !== null) {
    frontmatterObject[match[1]] = trimAndUnquote(match[2]);
    match = frontMatterItemRegex.exe(frontMatter);
}
```

**Comment by molant** [comment]
@sarvaje I've pushed my changes and addressed your feedback. Can you take another look?

**Comment by sarvaje** [comment]
@molant for me it is ok :)

**Comment by molant** [comment]
@ststimac @qzhou1607 I rebased with master to fix the conflicts. You might have to do a `pull --force` if you continue working on this branch.

**Pull request merged**